### PR TITLE
Add right-click object creation menu to viewport

### LINKED ISSUE #338: Right Mouse Button in the empty area of the viewport should bring up a context menu for adding a shape there.
This would be massively helpful for productivity.

### PR DESCRIPTION
**Pull request by DevPoodle**
Resolves #338

### CODE REVIEW

File: src/ui_parts/viewport.gd [hunk 1]
```diff
@@ -51,6 +52,22 @@ func center_frame() -> void:
 	adjust_view()
 	set_view((SVG.root_tag.get_size() - size / Indications.zoom) / 2)
 
+func create_object_context() -> Popup:
+	var btn_array: Array[Button] = []
+	var new_tid := PackedInt32Array([SVG.root_tag.get_child_count()])
+	var mouse_position := (get_mouse_position() / Indications.zoom + view.position).snapped(Vector2.ONE * GlobalSettings.save_data.snap)
```
Comment (MewPurPur) [inline src/ui_parts/viewport.gd:58]: Do the snapping only if it's enabled.

File: src/ui_parts/viewport.gd [hunk 2]
```diff
@@ -51,6 +52,22 @@ func center_frame() -> void:
...
+	
+	for tag in [TagPath.new(), TagCircle.new(), TagEllipse.new(), TagRect.new(), TagLine.new()]:
+		if tag is TagPath:
+			tag.attributes["d"]._commands.append(PathCommand.MoveCommand.new(0.0, 0.0, false))
```
Comment (MewPurPur) [inline src/ui_parts/viewport.gd:62]: This must use the setter for _commands.

File: src/ui_parts/viewport.gd [hunk 3]
```diff
@@ -51,6 +52,22 @@ func center_frame() -> void:
 	adjust_view()
 	set_view((SVG.root_tag.get_size() - size / Indications.zoom) / 2)
 
+func create_object_context() -> Popup:
+	var btn_array: Array[Button] = []
+	var new_tid := PackedInt32Array([SVG.root_tag.get_child_count()])
+	var mouse_position := (get_mouse_position() / Indications.zoom + view.position).snapped(Vector2.ONE * GlobalSettings.save_data.snap)
+	
+	for tag in [TagPath.new(), TagCircle.new(), TagEllipse.new(), TagRect.new(), TagLine.new()]:
```
Comment (MewPurPur) [inline src/ui_parts/viewport.gd:60]: Why do we need to create five tags when instantiating just a popup?

**Review by MewPurPur — COMMENTED** [review]
I think this code needs to be inside handles_manager.gd, and also work even if it's not visible

File: src/ui_parts/viewport.gd [hunk 4]
```diff
@@ -51,6 +52,22 @@ func center_frame() -> void:
...
+		if tag is TagPath:
+			tag.attributes["d"]._commands.append(PathCommand.MoveCommand.new(0.0, 0.0, false))
+		
+		tag.attributes["transform"].set_transform_list([AttributeTransform.TransformTranslate.new(mouse_position.x, mouse_position.y)] as Array[AttributeTransform.Transform])
```
Comment (MewPurPur) [inline src/ui_parts/viewport.gd:64]: I'd prefer if we do this on an attribute basis, i.e. changing cx and cy for circles.

File: src/ui_parts/viewport.gd [hunk 5]
```diff
@@ -51,6 +52,22 @@ func center_frame() -> void:
...
+		btn_array.append(Utils.create_btn(tr(tag.name), SVG.root_tag.add_tag.bind(tag, new_tid), false, tag.icon))
```
Comment (MewPurPur) [inline src/ui_parts/viewport.gd:65]: We don't need to use tr(tag.name) as tag names aren't translated right now, they are meant to be like the tags in code at the moment.

**Comment by MewPurPur** [comment]
Thank you for the initial work! I am working on a PR that will supersede this one.

**Comment by DevPoodle** [comment]
Cool! This is a feature I've been wanting for a while, so it'll be nice to have it regardless.

**Pull request closed without merging**